internal/controller/delayed_restart_manager_test.go

```diff
@@ -0,0 +1,808 @@
+	if !manager.IsScheduled(target) {
+		t.Error("IsScheduled() returned false after scheduling")
+	}
+
+	// Wait for execution with timeout
+	select {
+	case executedAt := <-executed:
+		elapsed := executedAt.Sub(startTime)
+		if elapsed < delay {
+			t.Errorf("Restart executed too early, elapsed=%v, expected at least %v", elapsed, delay)
+		}
+		if elapsed > delay+50*time.Millisecond {
+			t.Errorf("Restart executed too late, elapsed=%v, expected around %v", elapsed, delay)
+		}
+	case <-time.After(delay + 200*time.Millisecond):
+		t.Error("ScheduleRestart() did not execute within expected time")
+	}
+
+	// Check that task is removed after execution
+	time.Sleep(10 * time.Millisecond)
+	if manager.IsScheduled(target) {
+		t.Error("IsScheduled() returned true after execution")
+	}
+}
+
+//nolint:cyclop // Test function complexity is acceptable for comprehensive testing
+func TestDelayedRestartManager_ScheduleRestart_MultipleRulesHighestDelayWins(t *testing.T) {
+	t.Parallel()
+
+	logger := logr.Discard()
+	manager := NewDelayedRestartManager(logger)
+	defer manager.Shutdown()
+
+	ctx := context.Background()
+	target := TargetKey{
+		Kind:      "Deployment",
+		Name:      "test-deployment",
+		Namespace: "default",
+	}
+
+	executed := make(chan struct{})
+	var executionCount atomic.Int32
+	restartFunc := func(ctx context.Context) error {
+		executionCount.Add(1)
+		select {
+		case executed <- struct{}{}:
+		default:
+		}
+
+		return nil
+	}
+
+	// Schedule with delay 100ms
+	delay1 := 100 * time.Millisecond
+	finalDelay1, isNew1, err1 := manager.ScheduleRestart(ctx, target, delay1, restartFunc)
+	if err1 != nil {
+		t.Errorf("First ScheduleRestart() returned error: %v", err1)
+	}
+	if !isNew1 {
+		t.Error("First ScheduleRestart() should return isNew=true")
+	}
+	if finalDelay1 != delay1 {
+		t.Errorf("First ScheduleRestart() returned finalDelay=%v, want %v", finalDelay1, delay1)
+	}
+
+	// Schedule with higher delay 200ms - should reschedule
+	delay2 := 200 * time.Millisecond
+	finalDelay2, isNew2, err2 := manager.ScheduleRestart(ctx, target, delay2, restartFunc)
+	if err2 != nil {
+		t.Errorf("Second ScheduleRestart() returned error: %v", err2)
+	}
+	if !isNew2 {
+		t.Error("Second ScheduleRestart() with higher delay should return isNew=true (rescheduled)")
+	}
+	if finalDelay2 != delay2 {
+		t.Errorf("Second ScheduleRestart() returned finalDelay=%v, want %v", finalDelay2, delay2)
+	}
+
+	// Schedule with lower delay 50ms - should keep existing higher delay
+	delay3 := 50 * time.Millisecond
+	finalDelay3, isNew3, err3 := manager.ScheduleRestart(ctx, target, delay3, restartFunc)
+	if err3 != nil {
+		t.Errorf("Third ScheduleRestart() returned error: %v", err3)
+	}
+	if isNew3 {
+		t.Error("Third ScheduleRestart() with lower delay should return isNew=false")
+	}
+	if finalDelay3 != delay2 {
+		t.Errorf("Third ScheduleRestart() returned finalDelay=%v, want %v (existing higher delay)", finalDelay3, delay2)
+	}
+
+	// Wait for execution
+	select {
```

Comment:
This should not be either or. Executed needs to be true and the restart time should be within the specified limit.